adam/learner.py

```diff
@@ -102,3 +106,77 @@ def describe(
             return immutabledict(((memorized_description, 1.0),))
         else:
             return immutabledict()
+
+
+@attrs
+class SubsetLanguageLearner(
+    Generic[PerceptionT, LinguisticDescriptionT],
+    LanguageLearner[PerceptionT, LinguisticDescriptionT],
+):
+    """
+    An implementation of `LanguageLearner` for subset learning based approach for single object detection.
+    """
+
+    _descriptions_to_properties: Dict[
+        LinguisticDescriptionT, Set[PropertyPerception]
+    ] = attrib(init=False, default=Factory(dict))
+
+    def observe(
+        self, learning_example: LearningExample[PerceptionT, LinguisticDescriptionT]
+    ) -> None:
+        if len(learning_example.perception.frames) != 1:
+            raise RuntimeError("Subset learner can only handle single frames for now")
+
+        is_new_description = True
+        observed_linguistic_description = learning_example.linguistic_description
+        if isinstance(
+            learning_example.perception.frames[0], DevelopmentalPrimitivePerceptionFrame
```

Comment:
If you're going to refer to something more than once, pull it out into a variable.